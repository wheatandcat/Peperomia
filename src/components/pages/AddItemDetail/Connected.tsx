import React, { memo, useCallback, useState } from 'react';
import { Alert } from 'react-native';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import {
  useCreateItemDetailMutation,
  NewItem,
  NewItemDetail,
} from 'queries/api/index';
import Plain from './Plain';
import { Props as IndexProps } from './';

dayjs.extend(advancedFormat);

type Props = IndexProps & {
  date: string;
  itemId: string;
  priority: number;
  onCallback: () => void;
};

export type ConnectedType = {
  date: string;
  onDismiss: () => void;
  onSave: (item: NewItem) => void;
  onIcons: (kind: string) => void;
} & State;

type State = {
  selectedKind: string;
};

const Connected: React.FC<Props> = memo((props) => {
  const [state, setState] = useState<State>({ selectedKind: '' });
  const [
    createItemDetailMutation,
    createItemDetailMutationData,
  ] = useCreateItemDetailMutation({
    async onCompleted() {
      await props?.onCallback?.();

      props.navigation.goBack();
    },
    onError(err) {
      Alert.alert('保存に失敗しました', err.message);
    },
  });

  const onSave = useCallback(
    (item: NewItem) => {
      const itemDetail: NewItemDetail = {
        ...item,
        priority: props.priority,
        date: props.date,
        itemId: props.itemId,
      };

      const variables = {
        itemDetail,
      };

      createItemDetailMutation({ variables });
    },
    [createItemDetailMutation, props]
  );

  const onDismiss = useCallback(() => {
    props.navigation.goBack();
  }, [props.navigation]);

  const onIcons = useCallback(
    (kind: string) => {
      props.navigation.navigate('Icons', {
        kind,
        onSelectIcon: (selectedKind: string) => {
          setState((s) => ({
            ...s,
            selectedKind,
          }));
        },
      });
    },
    [props]
  );

  return (
    <Plain
      {...state}
      date={props.date}
      mutationData={createItemDetailMutationData}
      onDismiss={onDismiss}
      onSave={onSave}
      onIcons={onIcons}
    />
  );
});

export default Connected;
