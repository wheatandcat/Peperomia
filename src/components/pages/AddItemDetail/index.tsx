import React, { memo } from 'react';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { RootStackParamList } from 'lib/navigation';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import Connected from './Connected';

dayjs.extend(advancedFormat);

type ScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'AddItemDetail'
>;
type ScreenRouteProp = RouteProp<RootStackParamList, 'AddItemDetail'>;

export type Props = {
  navigation: ScreenNavigationProp;
  route: ScreenRouteProp;
};

const AddItemDetail: React.FC<Props> = memo((props) => {
  const date =
    props.route.params?.date || dayjs().format('YYYY-MM-DDT00:00:00');
  const itemId = props.route.params?.itemId || '';
  const priority = props.route.params?.priority || 0;
  const onCallback = props.route.params?.onCallback;

  return (
    <Connected
      {...props}
      date={date}
      itemId={itemId}
      priority={priority}
      onCallback={onCallback}
    />
  );
});

export default AddItemDetail;
