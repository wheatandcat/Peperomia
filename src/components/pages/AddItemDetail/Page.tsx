import React, { memo } from 'react';
import CreateCalendar from 'components/templates/CreateCalendar/Page';
import { QueryProps } from './Plain';
import { ConnectedType } from './Connected';

type Props = ConnectedType & {
  mutationData: QueryProps['mutationData'];
};

const AddItemDetailPage: React.FC<Props> = (props) => {
  return (
    <CreateCalendar
      {...props}
      loading={props.mutationData.loading}
      itemDetail={null}
    />
  );
};

export default memo(AddItemDetailPage);
