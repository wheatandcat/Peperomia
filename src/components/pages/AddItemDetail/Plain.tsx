import React from 'react';
import { CreateItemDetailMutationResult } from 'queries/api/index';
import { ConnectedType } from './Connected';
import Page from './Page';

export type QueryProps = {
  mutationData: CreateItemDetailMutationResult;
};

type Props = QueryProps & ConnectedType;

const Plain: React.FC<Props> = (props) => {
  return <Page {...props} onDismiss={props.onDismiss} onSave={props.onSave} />;
};

export default Plain;
