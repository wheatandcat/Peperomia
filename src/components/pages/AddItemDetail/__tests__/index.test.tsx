import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import AddItemDetail, { Props } from '../';

const propsData = (): Props =>
  ({
    navigation: {
      setParams: jest.fn(),
      navigate: jest.fn(),
    },
    route: {
      params: {
        date: '2020-01-01 00:00:00',
        itemId: 'test',
        priority: 1,
        onCallback: jest.fn(),
      },
    },
  } as any);

describe('components/pages/AddItemDetail/index.tsx', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<AddItemDetail {...propsData()} />);
  });

  it('正常にrenderすること', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
