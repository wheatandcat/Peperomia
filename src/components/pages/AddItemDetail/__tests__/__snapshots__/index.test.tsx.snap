// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/AddItemDetail/index.tsx 正常にrenderすること 1`] = `
<Memo()
  date="2020-01-01 00:00:00"
  itemId="test"
  navigation={
    Object {
      "navigate": [MockFunction],
      "setParams": [MockFunction],
    }
  }
  onCallback={[MockFunction]}
  priority={1}
  route={
    Object {
      "params": Object {
        "date": "2020-01-01 00:00:00",
        "itemId": "test",
        "onCallback": [MockFunction],
        "priority": 1,
      },
    }
  }
/>
`;
